// @kontent-ai/smart-link/commonjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyUpdateOnItem = exports.applyUpdateOnItemAndLoadLinkedItems = exports.buildComponentElementLink = exports.buildElementLink = exports.buildKontentLink = exports.KontentSmartLinkEvent = exports.default = void 0;
var sdk_1 = require("./sdk");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return sdk_1.default; } });
Object.defineProperty(exports, "KontentSmartLinkEvent", { enumerable: true, get: function () { return sdk_1.KontentSmartLinkEvent; } });
var link_1 = require("./utils/link");
Object.defineProperty(exports, "buildKontentLink", { enumerable: true, get: function () { return link_1.buildKontentLink; } });
Object.defineProperty(exports, "buildElementLink", { enumerable: true, get: function () { return link_1.buildElementLink; } });
Object.defineProperty(exports, "buildComponentElementLink", { enumerable: true, get: function () { return link_1.buildComponentElementLink; } });
var liveReload_1 = require("./utils/liveReload");
Object.defineProperty(exports, "applyUpdateOnItemAndLoadLinkedItems", { enumerable: true, get: function () { return liveReload_1.applyUpdateOnItemAndLoadLinkedItems; } });
Object.defineProperty(exports, "applyUpdateOnItem", { enumerable: true, get: function () { return liveReload_1.applyUpdateOnItem; } });
__exportStar(require("./models"), exports);
