// @kontent-ai/smart-link/commonjs/lib/IFrameCommunicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IFrameCommunicator = void 0;
const iframe_1 = require("../utils/iframe");
const EventManager_1 = require("./EventManager");
const IFrameCommunicatorTypes_1 = require("./IFrameCommunicatorTypes");
const createUuid_1 = require("../utils/createUuid");
const errors_1 = require("../utils/errors");
class IFrameCommunicator {
    constructor() {
        this.events = new EventManager_1.EventManager();
        this.callbacks = new Map();
        this.addMessageListener = (type, listener) => {
            this.events.on(type, listener);
        };
        this.removeMessageListener = (type, listener) => {
            this.events.off(type, listener);
        };
        this.sendMessageWithResponse = (type, data, callback, metadata) => {
            const requestId = (0, createUuid_1.createUuid)();
            this.registerCallback(requestId, callback);
            this.sendMessage(type, data, metadata, requestId);
        };
        this.sendMessage = (type, data, metadata, requestId) => {
            if (!(0, iframe_1.isInsideIFrame)()) {
                throw (0, errors_1.InvalidEnvironmentError)('IFrameCommunicator: iframe message can only be send while inside iframe.');
            }
            const message = { type, data, metadata, requestId };
            window.parent.postMessage(message, '*');
        };
        this.onMessage = (event) => {
            if (!event.data)
                return;
            const message = event.data;
            this.events.emit(message.type, message.data, message.metadata);
            if (message.requestId) {
                this.executeCallback(message.requestId, message.data);
            }
        };
        this.registerCallback = (requestId, callback) => {
            if (!callback) {
                return;
            }
            this.callbacks.set(requestId, callback);
        };
        this.executeCallback = (requestId, data) => {
            const callback = this.callbacks.get(requestId);
            if (callback) {
                callback(data);
                this.callbacks.delete(requestId);
            }
        };
    }
    initialize() {
        if (typeof window === 'undefined') {
            throw (0, errors_1.InvalidEnvironmentError)('IFrameCommunicator can only be initialized in a browser environment.');
        }
        window.addEventListener('message', this.onMessage, true);
    }
    destroy() {
        this.events.removeAllListeners();
        window.removeEventListener('message', this.onMessage, true);
    }
}
exports.IFrameCommunicator = IFrameCommunicator;
